aten/src/ATen/native/cuda/MemoryAccess.cuh

```diff
@@ -0,0 +1,259 @@
+    using backing_scalar_t = typename size_to_backing_type<scalar_size>::type1;
+    if (i == 0) {
+      v.x = typeless_cast<backing_scalar_t>(value);
+    } else {  // no boundary check here
+      v.y = typeless_cast<backing_scalar_t>(value);
+    }
+  }
+};
+
+template<int scalar_size>
+struct vector<scalar_size, 4> {
+  typename size_to_backing_type<scalar_size>::type4 v;
+
+  template<typename scalar_t>
+  __device__ inline scalar_t get(int i) {
+    switch (i) {
+    case 0:
+      return typeless_cast<scalar_t>(v.x);
+    case 1:
+      return typeless_cast<scalar_t>(v.y);
+    case 2:
+      return typeless_cast<scalar_t>(v.z);
+    }
+    return typeless_cast<scalar_t>(v.w); // no boundary check here
+  }
+
+  template<typename scalar_t>
+  __device__ inline void set(int i, scalar_t value) {
+    using backing_scalar_t = typename size_to_backing_type<scalar_size>::type1;
+    switch (i) {
+    case 0:
+      v.x = typeless_cast<backing_scalar_t>(value);
+      break;
+    case 1:
+      v.y = typeless_cast<backing_scalar_t>(value);
+      break;
+    case 2:
+      v.z = typeless_cast<backing_scalar_t>(value);
+      break;
+    default:  // no boundary check here
+      v.w = typeless_cast<backing_scalar_t>(value);
+    }
+  }
+};
+
+}  // namespace
+
+template <
+  int num_threads_,        // number of threads in a block.
+  int thread_work_size_    // number of elements each block needs to handle.
+>
+struct policies {
+
+  struct common {
+    static constexpr int num_threads = num_threads_;
+    static constexpr int thread_work_size = thread_work_size_;
+    static constexpr int block_work_size = thread_work_size_ * num_threads_;
+  };
+
+  struct checked_unroll : public common {
+
+    int remaining;
+
+    __device__ checked_unroll(int remaining): remaining(remaining) {}
+
+    template<typename accessor_t, typename scalar_t>
+    __device__ inline void load(accessor_t to, scalar_t *from) {
+      int thread_idx = threadIdx.x;
+      #pragma unroll
+      for (int i = 0; i < thread_work_size_; i++) {
+        if (thread_idx >= remaining) {
+          return;
+        }
+        to(i) = from[thread_idx];
+        thread_idx += num_threads_;
+      }
+    }
+
+    template<typename accessor_t, typename scalar_t>
+    __device__ inline void store(scalar_t *to, accessor_t from) {
+      int thread_idx = threadIdx.x;
+      #pragma unroll
+      for (int i = 0; i < thread_work_size_; i++) {
+        if (thread_idx >= remaining) {
+          return;
+        }
+        to[thread_idx] = from(i);
+        thread_idx += num_threads_;
+      }
+    }
+  };
+
+  // Functions here does not do boundary check. It assumes the whole block
+  // has its job to do. So the reminders should be handled by the the caller
+  // manually.
+
+  template <int vec_size>  // vec_size: number of scalars, can be 1, 2, or 3.
+  struct vectorized : public common {
+    static constexpr int loop_size = thread_work_size_ / vec_size;
```

Comment:
what happens if thread_work_size_ is not a multiple of vec_size? (it is for all instantiations currently, but I don't think there's a check for it).

Reply:
check added